shm/src/ship.cpp

```diff
@@ -0,0 +1,87 @@
+#include "ship.hpp"
+#include "locators.hpp"
+
+#include <algorithm>
+#include <iostream>
+
+Ship::Ship() : id_(-1) {}
+
+Ship::Ship(int capacity,
+           int maxCrew,
+           int speed,
+           const std::string& name,
+           size_t id,
+           Delegate* delegate)
+    : delegate_(delegate),
+      capacity_(capacity),
+      maxCrew_(maxCrew),
+      crew_(0),
+      speed_(speed),
+      name_(name),
+      id_(id) {
+    TimeServiceLocator::getTime()->addObserver(this);
+}
+
+Ship::Ship(int maxCrew, int speed, size_t id, Delegate* delegate)
+    : Ship(100, maxCrew, speed, "", id, delegate) {}
+
+void Ship::setName(const std::string& name) {
+    name_ = name;
+}
+
+Ship& Ship::operator-=(size_t crew) {
+    (crew_ < crew) ? crew_ = 0 : crew_ -= crew;
+    return *this;
+}
+Ship& Ship::operator+=(size_t crew) {
+    (crew_ + crew > maxCrew_) ? crew_ = maxCrew_ : crew_ += crew;
```

Comment:
I think it's better to compare crew_ and crew, because if we add crew_ and crew we risk going out of size_t range,which may result in gargabe value that can be lower than crew limit.